File: soungmunkim/C++/GraphSearch_BFS_DFS/1334_FindTheCity_FloydWarshall_ShortestPaths.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <climits>
#include <limits>
#include <string>

using namespace std;

class Vertex {
private:
    int id_;
    int key_ = 0;
    unordered_map<Vertex*, int> adjacency_list_;
    Vertex* previous_ = nullptr;

public:
    Vertex(int id) {
        this->id_ = id;
    }

    void AddEdge(Vertex* vertex, int weight) {
        adjacency_list_[vertex] = weight;
    }

    int GetId() { 
        return this->id_; 
    }

    void SetKey(int key) {
        this->key_ = key;
    }

    int GetKey() {
        return this->key_;
    }

    unordered_map<Vertex*, int>& GetAdjacencyList() { 
        return this->adjacency_list_; 
    }

    void SetPrevious(Vertex* prev) { 
        this->previous_ = prev; 
    }
    
    Vertex* GetPrevious() { 
        return this->previous_; 
    }
};

class Graph {
private:
    vector<Vertex*> vertices_;

public:
    Graph() {}

    ~Graph() {
        for (Vertex* vertex : this->vertices_) {
            delete vertex;
        }
        this->vertices_.clear();
    }

    Vertex* addNode(int id) {
        Vertex* newVertex = new Vertex(id);
        this->vertices_.push_back(newVertex);
        return newVertex;
    }

    void addEdge(Vertex* start, Vertex* end, int weight) {
        start->AddEdge(end, weight);
    }

    vector<Vertex*>& GetVertices() { 
        return this->vertices_; 
    }

    int GetNumVertices() {
        return this->vertices_.size();
    }
};

// G 그래프를 기반으로 인접 행렬을 초기화하는 함수
vector<vector<float>> initialize_adj_matrix(Graph* G) {
    // 그래프의 정점 수를 가져옵니다.
    int N = G->GetNumVertices();
    // 그래프의 모든 정점을 가져옵니다.
    vector<Vertex*> vertices = G->GetVertices();
    vector<vector<float>> W;
    // W를 NxN 크기로 확장하고 모든 값을 무한대로 초기화합니다.
    W.resize(N, vector<float>(N, numeric_limits<float>::infinity()));

    // 모든 정점 조합에 대해
    for (int i=0;i<N;i++) {
        for (int j=0;j<N;j++) {
            Vertex* u = vertices[i];
            Vertex* v = vertices[j];
            // 같은 정점이면 0으로 설정합니다.
            if (u == v) { W[i][j] = 0; }
            else {
                // u의 인접 리스트를 가져옵니다.
                auto adj = u->GetAdjacencyList();
                // v가 u의 인접 리스트에 있으면, 해당 가중치로 설정합니다.
                if (adj.find(v) != adj.end()) {
                    W[i][j] = adj[v];
                }
            }
        }
    }
    // 인접 행렬 W를 반환합니다.
    return W;
}

// 그래프 G에 대한 Floyd-Warshall 알고리즘을 실행하는 함수
pair<vector<vector<float>>, vector<vector<float>>> floyd_warshall(Graph* G) {
    // 그래프의 정점 수를 가져옵니다.
    int N = G->GetNumVertices();
    // 인접 행렬을 초기화합니다.
    vector<vector<float>> W = initialize_adj_matrix(G);
    
    vector<vector<float>> D;
    vector<vector<float>> P;
    // D와 P를 NxN 크기로 확장합니다.
    D.resize(N, vector<float>(N, 0));
    P.resize(N, vector<float>(N, numeric_limits<float>::infinity()));

    // D와 P를 초기 설정합니다.
    for (int i=0;i<N;i++) {
        for (int j=0;j<N;j++) {
            D[i][j] = W[i][j];
            // i와 j가 다르며, W[i][j]가 무한대가 아니면, P[i][j]를 i로 설정합니다.
            if (i != j && W[i][j] != numeric_limits<float>::infinity()) {
                P[i][j] = i;
            }
        }
    }

    // 모든 k에 대해 최단 경로를 업데이트합니다.
    for (int k=0;k<N;k++) {
        for (int i=0;i<N;i++) {
            for (int j=0;j<N;j++) {
                // 경로 i->k->j가 i->j보다 짧으면 업데이트합니다.
                if (D[i][k] + D[k][j] < D[i][j]) {
                    D[i][j] = D[i][k] + D[k][j];
                    P[i][j] = P[k][j];
                }
            }
        }
    }

    // 최단 거리 행렬 D와 경로 행렬 P를 반환합니다.
    return make_pair(D, P);
}

// 주어진 후속자 행렬과 시작 및 종료 인덱스를 사용하여 최단 경로를 생성하는 함수
vector<int> construct_path(vector<vector<float>>& pred, int start, int end) {
    vector<int> path;
    // 시작과 종료가 동일하거나 경로가 없는 경우
    if (start == end || pred[start][end] == -1) {
        path.push_back(start);
        if (start != end) path.push_back(end);
    } else {
        // 재귀적으로 경로를 구성
        vector<int> temp = construct_path(pred, start, pred[start][end]);
        // 임시 경로를 기본 경로에 추가
        path.insert(path.end(), temp.begin(), temp.end());
        path.push_back(end);
    }
    return path;
}

// 그래프 G와 시작 정점을 입력으로 받아 최단 경로를 생성하는 함수
unordered_map<Vertex*, vector<Vertex*>> floyd_warshall_shortest_paths(Graph* G, Vertex* start) {
    // floyd_warshall 알고리즘을 사용하여 최단 거리와 후속자 행렬을 가져옴
    auto result = floyd_warshall(G);
    vector<vector<float>> predecessors = result.second;
    vector<Vertex*> vertices = G->GetVertices();
    
    int start_idx = -1;
    // 시작 정점의 인덱스를 찾음
    for (int i = 0; i < vertices.size(); ++i) {
        if (vertices[i] == start) {
            start_idx = i;
            break;
        }
    }

    unordered_map<Vertex*, vector<Vertex*>> shortest_paths;
    // 각 목적지 정점에 대해
    for (int j = 0; j < vertices.size(); ++j) {
        // 최단 경로의 인덱스를 생성
        vector<int> id_paths = construct_path(predecessors, start_idx, j);
        vector<Vertex*> vertex_path;
        // 인덱스 경로를 정점 경로로 변환
        for (int id : id_paths) {
            vertex_path.push_back(vertices[id]);
        }
        // 결과 맵에 정점 경로 저장
        shortest_paths[vertices[j]] = vertex_path;
    }

    // 모든 최단 경로를 반환
    return shortest_paths;
}

// G 그래프에서 start 정점으로부터 모든 다른 정점까지의 최단 경로를 출력하는 함수
void PrintShortestPath(Graph* G, Vertex* start, int threshold) {
    // threshold가 필요 없는 경우, 함수 인자에서 제거하면 됩니다

    // 그래프 G에서 모든 정점을 가져옴
    vector<Vertex*> vertices = G->GetVertices();

    // floyd_warshall 알고리즘을 사용하여 모든 쌍의 최단 거리를 계산
    vector<vector<float>> distance = floyd_warshall(G).first;

    // 시작 정점에서 모든 다른 정점까지의 최단 경로를 계산
    unordered_map<Vertex*, vector<Vertex*>> shortest_paths = floyd_warshall_shortest_paths(G, start);

    int start_idx = -1;
    // 시작 정점의 인덱스를 찾음
    for (int i = 0; i < vertices.size(); ++i) {
        if (vertices[i] == start) {
            start_idx = i;
            break;
        }
    }
    
    // 모든 최단 경로에 대해
    for (auto& item : shortest_paths) {
        Vertex* to = item.first;  // 목적지 정점
        vector<Vertex*> path = item.second;  // 시작에서 목적지까지의 경로

        // 시작 정점과 목적지 정점이 동일한 경우는 건너뛴다
        if (start == to) { continue; }

        // threshold에 따라서 출력 여부를 결정하는 부분 (필요하지 않은 경우 주석 처리 가능)
        int to_idx = -1;
        for (int i = 0; i < vertices.size(); ++i) {
            if (vertices[i] == to) {
                to_idx = i;
                break;
            }
        }

        // 거리가 임계값을 초과하면 해당 경로를 건너뜀
        if (distance[start_idx][to_idx] > threshold) { continue; }

        // 시작 정점과 목적지 정점의 ID를 출력
        cout << "(" << start->GetId() << ", " << to->GetId() << ") : ";

        // 경로 상의 중간 정점들을 출력
        bool is_start = true;
        cout << "[";
        for (Vertex* intermediate : path) {
            if (!is_start) {
                cout << ", ";
            }
            cout << intermediate->GetId();
            is_start = false;
        }
        // 최단 거리를 출력
        cout << "] distance : " << distance[start_idx][to_idx] << endl;
    }
}


int main() {
    Graph* graph = new Graph();

    Vertex* node0 = graph->addNode(0);
    Vertex* node1 = graph->addNode(1);
    Vertex* node2 = graph->addNode(2);
    Vertex* node3 = graph->addNode(3);

    graph->addEdge(node0, node1, 3);
    graph->addEdge(node1, node2, 1);
    graph->addEdge(node1, node3, 4);
    graph->addEdge(node1, node0, 3);
    graph->addEdge(node2, node1, 1);
    graph->addEdge(node2, node3, 1);
    graph->addEdge(node3, node1, 4);
    graph->addEdge(node3, node2, 1);

    int threshold = 4;
    for (Vertex* v : graph->GetVertices()) {
        cout << "Shortest path from " << v->GetId() << " : " << endl;
        PrintShortestPath(graph, v, threshold);
        cout << "=========================== DONE ===========================" << endl;
    }
    
    return 0;
}

File: sungwoopark95/python/Graph/1334_FindTheCity_FloydWarshall_ShortestPaths.py
##################### Floyd Warshall 알고리즘으로 shortest path 찾고 해당 path 출력 #####################
#################################### Vertex, Graph class 버전 #####################################
from typing import Dict, List, Tuple
import math

class Vertex:
    def __init__(self, id_: int):
        self.id_ = id_  # 정점의 ID
        self.key_ = 0  # 키 (예: 프림 알고리즘에서 사용됨)
        self.adjacency_list_ = {}  # 인접한 정점과 가중치를 담는 목록
        self.previous_ = None  # 알고리즘에서 사용될 이전 정점 참조

    def add_edge(self, vertex, weight):
        self.adjacency_list_[vertex] = weight

    def get_id(self):
        return self.id_

    def set_key(self, key):
        self.key_ = key

    def get_key(self):
        return self.key_

    def get_adjacency_list(self):
        return self.adjacency_list_

    def set_previous(self, prev):
        self.previous_ = prev

    def get_previous(self):
        return self.previous_

class Graph:
    def __init__(self):
        self.vertices_ = []  # 그래프 내의 모든 정점들

    def __del__(self):
        self.vertices_.clear()

    def add_node(self, id_: int) -> Vertex:
        new_vertex = Vertex(id_)
        self.vertices_.append(new_vertex)
        return new_vertex

    def add_edge(self, start: Vertex, end: Vertex, weight: int):
        start.add_edge(end, weight)

    def get_vertices(self):
        return self.vertices_

    def get_num_vertices(self):
        return len(self.vertices_)

##################################### 인접행렬 초기화하는 함수 #####################################
def initialize_adj_matrix(G: Graph) -> List[List[float]]:
    N = G.get_num_vertices()
    vertices = G.get_vertices()
    W = [[math.inf for _ in range(N)] for _ in range(N)]

    for i in range(N):
        for j in range(N):
            u, v = vertices[i], vertices[j]
            if u == v:
                W[i][j] = 0
            elif v in u.get_adjacency_list():
                W[i][j] = u.get_adjacency_list()[v]
    return W

################################## Floyd Warshall 수행하는 함수 ##################################
################################# Distance, Parent Matrix 반환 #################################
def floyd_warshall(G: Graph) -> Tuple[List[List[float]], List[List[float]]]:
    N = G.get_num_vertices()
    W = initialize_adj_matrix(G)
    D = [[0 for _ in range(N)] for _ in range(N)]
    P = [[math.inf for _ in range(N)] for _ in range(N)]
    
    # Initialize D and P matrices
    for i in range(N):
        for j in range(N):
            D[i][j] = W[i][j]
            if i != j and W[i][j] != math.inf:
                P[i][j] = i

    # Update matrices
    for k in range(N):
        for i in range(N):
            for j in range(N):
                if D[i][k] + D[k][j] < D[i][j]:
                    D[i][j] = D[i][k] + D[k][j]
                    P[i][j] = P[k][j]
    return D, P

##################### Parent Matrix, start index, end index로 경로 만드는 함수 ######################
def construct_path(pred: List[List[float]], start: int, end: int) -> List[int]:
    path = []
    if start == end or pred[start][end] == math.inf:
        path.extend([start, end])
    else:
        path.extend(construct_path(pred, start, pred[start][end]))
        path.append(end)
    return path

########################### start point에서 각각의 point로 가는 최단 경로 반환 ###########################
def floyd_warshall_shortest_paths(G: Graph, start: Vertex) -> Dict[Vertex, List[Vertex]]:
    _, P = floyd_warshall(G)
    vertices = G.get_vertices()
    start_idx = vertices.index(start)

    shortest_paths = {}
    for j in range(len(vertices)):
        id_path = construct_path(P, start_idx, j)
        vertex_path = [vertices[id_] for id_ in id_path]
        shortest_paths[vertices[j]] = vertex_path

    return shortest_paths

########################### start point에서 각각의 point로 가는 최단 경로 출력 ###########################
def print_shortest_path(G: Graph, start: Vertex, threshold: int):
    vertices = G.get_vertices()
    distance = floyd_warshall(G)[0]
    shortest_paths = floyd_warshall_shortest_paths(G, start)

    start_idx = vertices.index(start)
    
    for to, path in shortest_paths.items():
        if start == to:
            continue
        
        to_idx = vertices.index(to)
        
        if distance[start_idx][to_idx] > threshold:
            continue
        
        path_str = "->".join([str(v.get_id()) for v in path])
        print(f"({start.get_id()}, {to.get_id()}) : [{path_str}]")

